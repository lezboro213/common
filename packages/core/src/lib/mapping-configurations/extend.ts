import type {
    Dictionary,
    Mapping,
    MappingConfiguration,
    MappingProperty,
    ModelIdentifier,
} from '../types';
import { MappingClassId } from '../types';
import { getMapping } from '../utils/get-mapping';
import { isSamePath } from '../utils/is-same-path';

export function extend<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TExtendSource extends Dictionary<TExtendSource>,
    TExtendDestination extends Dictionary<TExtendDestination>
>(
    mapping: Mapping<TExtendSource, TExtendDestination>
): MappingConfiguration<TSource, TDestination>;
export function extend<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TExtendSource extends Dictionary<TExtendSource>,
    TExtendDestination extends Dictionary<TExtendDestination>
>(
    source: ModelIdentifier<TExtendSource>,
    destination: ModelIdentifier<TExtendDestination>
): MappingConfiguration<TSource, TDestination>;
export function extend<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TExtendSource extends Dictionary<TExtendSource>,
    TExtendDestination extends Dictionary<TExtendDestination>
>(
    sourceOrMapping:
        | ModelIdentifier<TExtendSource>
        | Mapping<TExtendSource, TExtendDestination>,
    destination?: ModelIdentifier<TExtendDestination>
): MappingConfiguration<TSource, TDestination> {
    return (mapping) => {
        let mappingToExtend: Mapping<TExtendSource, TExtendDestination>;
        if (Array.isArray(sourceOrMapping)) {
            mappingToExtend = sourceOrMapping as Mapping<
                TExtendSource,
                TExtendDestination
            >;
        } else {
            const mapper = mapping[MappingClassId.mapper];
            mappingToExtend = getMapping(mapper, sourceOrMapping, destination!);
        }

        const propsToExtend = mappingToExtend[MappingClassId.properties];

        for (let i = 0, length = propsToExtend.length; i < length; i++) {
            const [
                propToExtendKey,
                propToExtendMappingProp,
                propToExtendNestedMapping,
            ] = propsToExtend[i];
            const existProp = mapping[MappingClassId.customProperties].find(
                ([pKey]) => isSamePath(pKey, propToExtendKey)
            );
            if (existProp) continue;
            mapping[MappingClassId.customProperties].push([
                propToExtendKey,
                propToExtendMappingProp as MappingProperty<
                    TSource,
                    TDestination
                >,
                propToExtendNestedMapping,
            ]);
        }
    };
}
