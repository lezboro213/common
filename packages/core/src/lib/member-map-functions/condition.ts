import type {
    ConditionPredicate,
    ConditionReturn,
    Dictionary,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';
import { get } from '../utils/get';

export function condition<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    predicate: ConditionPredicate<TSource>,
    defaultValue?: TSelectorReturn
): ConditionReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.Condition,
        (source, sourceMemberPaths) => {
            if (predicate(source)) {
                return get(source, sourceMemberPaths) as TSelectorReturn;
            }

            return defaultValue as TSelectorReturn;
        },
    ];
}
