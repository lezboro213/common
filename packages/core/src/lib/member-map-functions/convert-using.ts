import type {
    Converter,
    ConvertUsingReturn,
    Dictionary,
    Selector,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';

export function convertUsing<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>,
    TConvertSourceReturn = SelectorReturn<TSource>
>(
    converter: Converter<TConvertSourceReturn, TSelectorReturn>,
    selector: Selector<TSource, TConvertSourceReturn>
): ConvertUsingReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.ConvertUsing,
        (source) => converter.convert(selector(source)),
    ];
}
