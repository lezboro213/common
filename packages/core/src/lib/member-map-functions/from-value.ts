import type { Dictionary, FromValueReturn, SelectorReturn } from '../types';
import { TransformationType } from '../types';

export function fromValue<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    rawValue: TSelectorReturn
): FromValueReturn<TSource, TDestination, TSelectorReturn> {
    return [TransformationType.FromValue, () => rawValue];
}
