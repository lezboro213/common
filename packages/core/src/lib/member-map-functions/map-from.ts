import type {
    Dictionary,
    MapFromReturn,
    Resolver,
    SelectorReturn,
    ValueSelector,
} from '../types';
import { TransformationType } from '../types';
import { isResolver } from '../utils/is-resolver';

export function mapFrom<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    from:
        | ValueSelector<TSource, TDestination, TSelectorReturn>
        | Resolver<TSource, TDestination, TSelectorReturn>
): MapFromReturn<TSource, TDestination, TSelectorReturn> {
    if (isResolver(from)) {
        return [TransformationType.MapFrom, from.resolve.bind(from)];
    }

    return [TransformationType.MapFrom, from];
}
