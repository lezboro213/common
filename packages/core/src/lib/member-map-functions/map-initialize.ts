import type { Dictionary, MapInitializeReturn, SelectorReturn } from '../types';
import { TransformationType } from '../types';
import { get } from '../utils/get';

export function mapInitialize<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    sourcePath: string[]
): MapInitializeReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.MapInitialize,
        (source) => get(source, sourcePath) as TSelectorReturn,
    ];
}
