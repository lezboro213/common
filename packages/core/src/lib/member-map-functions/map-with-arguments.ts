import type {
    Dictionary,
    MapWithArgumentsReturn,
    Resolver,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';
import { isResolver } from '../utils/is-resolver';

export function mapWithArguments<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    withArgumentsResolver:
        | ((
              source: TSource,
              extraArguments: Record<string, unknown>
          ) => TSelectorReturn)
        | Resolver<TSource, Record<string, unknown>, TSelectorReturn>
): MapWithArgumentsReturn<TSource, TDestination, TSelectorReturn> {
    if (isResolver(withArgumentsResolver)) {
        return [
            TransformationType.MapWithArguments,
            withArgumentsResolver.resolve.bind(withArgumentsResolver),
        ];
    }
    return [TransformationType.MapWithArguments, withArgumentsResolver];
}
