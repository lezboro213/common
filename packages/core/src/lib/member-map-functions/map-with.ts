import type {
    Dictionary,
    MapOptions,
    MapWithReturn,
    ModelIdentifier,
    SelectorReturn,
} from '../types';
import { TransformationType, ValueSelector } from '../types';

type Constructor<TModel> = new (...args: unknown[]) => TModel;

export function mapWith<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>,
    TWithDestination extends ModelIdentifier = ModelIdentifier,
    TWithSource extends ModelIdentifier = ModelIdentifier,
    TWithSourceValue extends ValueSelector = TWithSource extends Constructor<
        infer InferredWithSource
    >
        ? ValueSelector<TSource, InferredWithSource>
        : ValueSelector<TSource>
>(
    withDestination: TWithDestination,
    withSource: TWithSource,
    withSourceValue: TWithSourceValue
): MapWithReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.MapWith,
        (source, mapper, options) => {
            const sourceValue = withSourceValue(source);

            if (Array.isArray(sourceValue)) {
                return mapper.mapArray(
                    sourceValue,
                    withSource,
                    withDestination,
                    options as unknown as MapOptions<TSource[], TDestination[]>
                ) as unknown as TSelectorReturn;
            }

            return mapper.map(
                sourceValue,
                withSource,
                withDestination,
                options
            ) as unknown as TSelectorReturn;
        },
    ];
}
