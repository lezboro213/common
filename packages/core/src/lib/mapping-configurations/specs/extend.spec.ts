import { mapInitialize } from '../../member-map-functions/map-initialize';
import type { Mapping } from '../../types';
import { MappingClassId } from '../../types';
import { extend } from '../extend';

function getMappingProperty(path = 'foo') {
    return [
        [path],
        [[path], [mapInitialize([path])]],
        [String, String],
    ] as unknown as Mapping[MappingClassId.properties][number];
}

describe(extend.name, () => {
    it('should extend another mapping', () => {
        const mapping = [] as unknown as Mapping;
        mapping[MappingClassId.customProperties] = [];
        const mappingToExtend = [] as unknown as Mapping;
        mappingToExtend[MappingClassId.properties] = [getMappingProperty()];

        extend(mappingToExtend)(mapping);
        expect(mapping[MappingClassId.customProperties]).toMatchSnapshot();
    });

    it('should skip existing mapping properties on extended mapping', () => {
        const mapping = [] as unknown as Mapping;
        mapping[MappingClassId.customProperties] = [getMappingProperty()];
        const mappingToExtend = [] as unknown as Mapping;
        mappingToExtend[MappingClassId.properties] = [
            getMappingProperty(),
            getMappingProperty('bar'),
        ];

        extend(mappingToExtend)(mapping);
        expect(mapping[MappingClassId.customProperties]).toMatchSnapshot();
    });
});
