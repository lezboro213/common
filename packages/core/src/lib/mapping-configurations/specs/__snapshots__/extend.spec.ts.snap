// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extend should extend another mapping 1`] = `
Array [
  Array [
    Array [
      "foo",
    ],
    Array [
      Array [
        "foo",
      ],
      Array [
        Array [
          6,
          [Function],
        ],
      ],
    ],
    Array [
      [Function],
      [Function],
    ],
  ],
]
`;

exports[`extend should skip existing mapping properties on extended mapping 1`] = `
Array [
  Array [
    Array [
      "foo",
    ],
    Array [
      Array [
        "foo",
      ],
      Array [
        Array [
          6,
          [Function],
        ],
      ],
    ],
    Array [
      [Function],
      [Function],
    ],
  ],
  Array [
    Array [
      "bar",
    ],
    Array [
      Array [
        "bar",
      ],
      Array [
        Array [
          6,
          [Function],
        ],
      ],
    ],
    Array [
      [Function],
      [Function],
    ],
  ],
]
`;
